Docker/dagster/Dockerfile_user_code

```diff
@@ -1,10 +1,39 @@
-FROM python:3.12-slim
+# Copyright 2025 Lincoln Institute of Land Policy
+# SPDX-License-Identifier: Apache-2.0
 
-# Base requirements
-COPY requirements.txt user_code_requirements.txt
-RUN pip install -r user_code_requirements.txt
+# Based on: https://github.com/astral-sh/uv-docker-example/blob/main/multistage.Dockerfile
+
+# First, build the application in the `/app` directory.
+FROM ghcr.io/astral-sh/uv:python3.12-bookworm-slim AS builder
+ENV UV_COMPILE_BYTECODE=1 UV_LINK_MODE=copy
+
+# Disable Python downloads, because we want to use the system interpreter
+# across both images. If using a managed Python version, it needs to be
+# copied from the build image into the final image; see `standalone.Dockerfile`
+# for an example.
+ENV UV_PYTHON_DOWNLOADS=0
+
+WORKDIR /app
+RUN --mount=type=cache,target=/root/.cache/uv \
+    --mount=type=bind,source=uv.lock,target=uv.lock \
+    --mount=type=bind,source=pyproject.toml,target=pyproject.toml \
+    uv sync --frozen --no-install-project --no-dev
+ADD . /app
+RUN --mount=type=cache,target=/root/.cache/uv \
+    uv sync --frozen --no-dev
```

Comment:
What are these caches tied to? I know there were issues with using cache in Cloud Build but I guess images are just built on the vm..?